plugins/module_utils/_apypie.py

```diff
@@ -921,3 +921,188 @@ def path_with_params(self, params=None):
+        return cls(msg=msg, error=error)
+
+
+class ForemanApi(Api):
+
+    def __init__(self, **kwargs):
+        self.task_timeout = kwargs.pop('task_timeout', 60)
+        self.task_poll = 4
+        kwargs['api_version'] = 2
+        return super(ForemanApi, self).__init__(**kwargs)
+
+    def _resource(self, resource):
+        if resource not in self.resources:
+            raise Exception("The server doesn't know about {0}, is the right plugin installed?".format(resource))
+        return self.resource(resource)
+
+    def _resource_call(self, resource, *args, **kwargs):
+        return self._resource(resource).call(*args, **kwargs)
+
+    def _resource_prepare_params(self, resource, action, params):
+        api_action = self._resource(resource).action(action)
+        return api_action.prepare_params(params)
+
+    def resource_action(self, resource, action, params, options=None, data=None, files=None,
+                        ignore_task_errors=False):
+        resource_payload = self._resource_prepare_params(resource, action, params)
+        if options is None:
+            options = {}
+        try:
+            result = self._resource_call(resource, action, resource_payload, options=options, data=data, files=files)
+            is_foreman_task = isinstance(result, dict) and 'action' in result and 'state' in result and 'started_at' in result
+            if is_foreman_task:
+                result = self.wait_for_task(result, ignore_errors=ignore_task_errors)
+        except Exception as e:
+            msg = 'Error while performing {0} on {1}: {2}'.format(
+                action, resource, str(e))
+            raise ForemanApiException.from_exception(e, msg) from e
+        return result
+
+    def wait_for_task(self, task, ignore_errors=False):
+        duration = self.task_timeout
+        while task['state'] not in ['paused', 'stopped']:
+            duration -= self.task_poll
+            if duration <= 0:
+                raise ForemanApiException(msg="Timeout waiting for Task {0}".format(task['id']))
+            time.sleep(self.task_poll)
+
+            resource_payload = self._resource_prepare_params('foreman_tasks', 'show', {'id': task['id']})
+            task = self._resource_call('foreman_tasks', 'show', resource_payload)
+        if not ignore_errors and task['result'] != 'success':
+            msg = 'Task {0}({1}) did not succeed. Task information: {2}'.format(task['action'], task['id'], task['humanized']['errors'])
+            raise ForemanApiException(msg=msg)
+        return task
+
+    def show(self, resource, resource_id, params=None):
+        """
+        Execute the ``show`` action on an entity.
+
+        :param resource: Plural name of the api resource to show
+        :type resource: str
+        :param resource_id: The ID of the entity to show
+        :type resource_id: int
+        :param params: Lookup parameters (i.e. parent_id for nested entities)
+        :type params: Union[dict,None], optional
```

Comment:
There's also `Optional[dict]` according to https://docs.python.org/3/library/typing.html#typing.Optional so is this also sufficient?
```suggestion
        :type params: Optional[dict]
```

Reply:
Yeah, that's the same.